packages/gestalt/src/MasonryV2.js

```diff
@@ -0,0 +1,654 @@
+   *
+   * This is an experimental prop and may be removed in the future.
+   */
+  _logTwoColWhitespace?: (number) => void,
+  /**
+   * Experimental prop to measure all items in one batch
+   */
+  _measureAll?: boolean,
+};
+
+type MasonryRef = {
+  +handleResize: () => void,
+  +reflow: () => void,
+  +state: {
+    width: ?number,
+  },
+};
+
+/**
+ * [Masonry](https://gestalt.pinterest.systems/web/masonry) creates a deterministic grid layout, positioning items based on available vertical space. It contains performance optimizations like virtualization and support for infinite scrolling.
+ *
+ * ![Masonry light mode](https://raw.githubusercontent.com/pinterest/gestalt/master/playwright/visual-test/Masonry.spec.mjs-snapshots/Masonry-chromium-darwin.png)
+ *
+ */
+
+function createMeasurementStore<T1: { ... }, T2>(): MeasurementStore<T1, T2> {
+  return new MeasurementStore();
+}
+
+function subscribeToResizeEvent(callback: () => void) {
+  window.addEventListener('resize', callback);
+  return () => {
+    window.removeEventListener('resize', callback);
+  };
+}
+
+// helper hook to force update a component
+function useForceUpdate() {
+  const [, forceUpdate] = useReducer<number, void>((x) => x + 1, 0);
+  return forceUpdate;
+}
+
+function useElementWidth(element: ?HTMLDivElement) {
+  const getElementWidth = useCallback(() => {
+    const elWidth = element?.clientWidth;
+    return elWidth;
+  }, [element]);
+
+  const width = useSyncExternalStore(subscribeToResizeEvent, getElementWidth, () => null);
+  return width;
+}
+
+function useScrollContainer({
+  gridWrapper,
+  scrollContainer,
+}: {
+  gridWrapper: ?HTMLElement,
+  scrollContainer: ?HTMLElement,
+}) {
+  const [containerHeight, setContainerHeight] = useState(0);
+  const [containerOffset, setContainerOffset] = useState(0);
+  const subscribeToScrollEvent = useCallback(
+    (callback: () => void) => {
+      scrollContainer?.addEventListener('scroll', callback);
+      return () => {
+        scrollContainer?.removeEventListener('scroll', callback);
+      };
+    },
+    [scrollContainer],
+  );
+
+  const scrollTop = useSyncExternalStore(
+    subscribeToScrollEvent,
+    () => (scrollContainer ? getScrollPos(scrollContainer) : 0),
+    () => 0,
+  );
+
+  useIsomorphicLayoutEffect(() => {
+    if (scrollContainer) {
+      setContainerHeight(getElementHeight(scrollContainer));
+    }
+  }, [scrollContainer]);
+
+  useIsomorphicLayoutEffect(() => {
+    // recalculate container offset right before browser paints in order to ensure getBoundingClientRect is accurate
+    if (scrollContainer && gridWrapper instanceof HTMLElement) {
+      const relativeScrollTop = getRelativeScrollTop(scrollContainer);
+      setContainerOffset(gridWrapper.getBoundingClientRect().top + relativeScrollTop);
+    }
+  }, [gridWrapper, scrollContainer]);
+
+  return {
+    containerHeight,
+    containerOffset,
+    scrollTop,
+  };
+}
+
+function useFetchOnScroll({
```

Comment:
FYI: this functionality mostly lives in the `FetchItems` component in current Masonry